loopy/kernel/dependency.py

```diff
@@ -0,0 +1,208 @@
+    write_maps: list[AccessRelation] = [AccessRelation(insn.id, var, 
+                                                      get_map(var, insn),
+                                                      AccessType.WRITE)
+                 for insn in knl.instructions
+                 for var in write_var_list(insn)]
+
+    write_read: list[HappensBefore] = [HappensBefore(read.id,
+                                        write.variable_name,
+                                        get_dependency_relation(write.relation,
+                                                                read.relation),
+                                        DependencyType.WRITE_READ)
+                                       for write in write_maps
+                                       for read in read_maps
+                                       if write.variable_name == read.variable_name]
+    read_write: list[HappensBefore] = [HappensBefore(write.id,
+                                        read.variable_name,
+                                        get_dependency_relation(read.relation,
+                                                                write.relation),
+                                        DependencyType.READ_WRITE)
+                                       for read in read_maps
+                                       for write in write_maps
+                                       if read.variable_name == write.variable_name]
+    write_write: list[HappensBefore] = [HappensBefore(write2.id,
+                                         write1.variable_name,
+                                         get_dependency_relation(write1.relation,
+                                         write2.relation),
+                                         DependencyType.WRITE_WRITE)
+                                        for write1 in write_maps
+                                        for write2 in write_maps
+                                        if write1.variable_name == write2.variable_name]
+
+    return write_read, read_write, write_write
+
+def generate_execution_order(knl: LoopKernel) -> frozenset[isl.Map]:
+    """Generate the "happens-before" execution order that *must* be respected by
+    any transformation. Calls :function:`generate_dependency_relations` to get
+    the information needed to compute the execution order.
+
+    :arg knl: A :class:`loopy.LoopKernel` containing the instructions for which
+    to generate a "happens-before" execution order.
+
+    :returns: A :class:`frozenset` of :class:`isl.Map` representing the
+    execution required by the dependencies in a loopy program.
+    """
+
+    write_read, read_write, write_write = generate_dependency_relations(knl)
+    
+    execution_order: frozenset[isl.Map] = frozenset()
+
+    for insn in knl.instructions:
+        domain: isl.BasicSet = knl.get_inames_domain(insn.within_inames)
+        insn_order: isl.Map = domain.lex_lt_set(domain)
```

Comment:
This shouldn't start from lexicographic, but rather from an already-existing happens-before.